w3af/plugins/audit/open_api_auth.py

```diff
@@ -0,0 +1,372 @@
+        # The check works only with API endpoints.
+        # crawl.open_api plugin has to be called before.
+        if not self._is_api_spec_available():
+            om.out.information("Could not find an API specification, skip")
+            return
+
+        # The API spec must define authentication mechanisms.
+        if not self._has_security_definitions_in_spec():
+            om.out.information("The API spec has no security definitions, skip")
+            return
+
+        # The check only runs if the API specification
+        # requires authentication for a endpoint.
+        if not self._has_auth(freq):
+            om.out.information("Request doesn't have auth info, skip")
+            return
+
+        # Remove auth info from the request.
+        freq_with_no_auth = self._remove_auth(freq)
+
+        # Send the request to the server, and check if access denied.
+        self._uri_opener.send_mutant(freq_with_no_auth,
+                                     callback=self._check_response_code,
+                                     debugging_id=debugging_id)
+
+    def _check_response_code(self, freq, response):
+        """
+        Check if an HTTP request contains an expected response code.
+
+        :param freq: A FuzzableRequest to the API endpoint.
+        :param response: The HTTP response associated with the fuzzable request.
+        """
+        if response.get_code() not in self._expected_codes:
+            desc = 'A %s request without auth info was send to %s. ' \
+                   'The server replied with an unexpected code (%d), expected %s' \
+                   % (freq.get_method(), freq.get_url(),
+                      response.get_code(), self._expected_codes)
+
+            # Should severity depend on response codes?
+            # For example, 2xx codes results to HIGH, but 4xx may be MEDIUM/LOW
+            v = Vuln.from_fr('Broken authentication', desc,
+                             severity.HIGH, response.id,
+                             self.get_name(), freq)
+
+            v['response_code'] = response.get_code()
+            v['method'] = freq.get_method()
+
+            self.kb_append_uniq(self, 'open_api_auth', v)
+
+    def _has_security_definitions_in_spec(self):
+        """
+        :return: True if the API spec contains 'securityDefinitions' section,
+                 False otherwise.
+        """
+        if self._spec.security_definitions:
+            return True
+
+        return False
+
+    def _is_api_spec_available(self):
+        """
+        Make sure that we have API specification.
+        The API spec has to be provided by crawl.open_api plugin
+        which should be called before.
+
+        The plugins use the global knowledge base to share the API spec.
+
+        :return: True if API specification is available, False otherwise.
+        """
+        if self._spec:
+            return True
+
+        specification_handler = kb.kb.raw_read('open_api', 'specification_handler')
```

Comment:
What if there is more than one openapi document found during a scan?

Maybe you should store a list and call it `kb.kb.raw_read('open_api', 'specification_handlers')` (added an `s` at the end)

This also means that you'll have to touch other areas of this plugin to process all specification handlers, not just one.